src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -53,40 +53,53 @@ impl<F: FieldExt + PrimeFieldBits, const K: usize> LookupRangeCheckConfig<F, K>
         meta.enable_equality(running_sum.into());
 
         let q_lookup = meta.complex_selector();
-        let q_lookup_short = meta.complex_selector();
-        let q_lookup_bitshift = meta.selector();
+        let q_running = meta.complex_selector();
+        let q_bitshift = meta.selector();
         let config = LookupRangeCheckConfig {
             q_lookup,
-            q_lookup_short,
-            q_lookup_bitshift,
+            q_running,
+            q_bitshift,
             running_sum,
             table_idx,
             _marker: PhantomData,
         };
 
-        // Lookup for range checks K bits and above.
         meta.lookup(|meta| {
             let q_lookup = meta.query_selector(config.q_lookup);
+            let q_running = meta.query_selector(config.q_running);
             let z_cur = meta.query_advice(config.running_sum, Rotation::cur());
-            let z_next = meta.query_advice(config.running_sum, Rotation::next());
+
+            // In the case of a running sum decomposition, we recover the word from
+            // the difference of the running sums:
             //    z_i = 2^{K}⋅z_{i + 1} + a_i
             // => a_i = z_i - 2^{K}⋅z_{i + 1}
-            let word = z_cur - z_next * F::from_u64(1 << K);
+            let running_sum_lookup = {
+                let running_sum_word = {
+                    let z_next = meta.query_advice(config.running_sum, Rotation::next());
+                    z_cur.clone() - z_next * F::from_u64(1 << K)
+                };
 
-            vec![(q_lookup * word, config.table_idx)]
-        });
+                q_running.clone() * running_sum_word
+            };
 
-        // Lookup used in range checks up to S bits, where S < K.
-        meta.lookup(|meta| {
-            let q_lookup_short = meta.query_selector(config.q_lookup_short);
-            let word = meta.query_advice(config.running_sum, Rotation::cur());
+            // In the short range check, the word is directly witnessed.
+            let short_lookup = {
+                let short_word = z_cur;
+                let q_short = Expression::Constant(F::one()) - q_running;
+
+                q_short * short_word
+            };
 
-            vec![(q_lookup_short * word, config.table_idx)]
+            // Combine the running sum and short lookups:
+            vec![(
+                q_lookup * (running_sum_lookup + short_lookup),
```

Comment:
If the item to be looked up were always in the `word` column and `z` were in a separate column, then it would be possible to make `q_running` a simple rather than complex selector. This can be a later optimization.